File: editor/api/controlAgent.ts
import client from "@/client";

import {
  GET_CONTROL_MAP,
  GET_CONTROL_RECORD,
  EDIT_CONTROL_FRAME,
  EDIT_CONTROL_FRAME_TIME,
  DELETE_CONTROL_FRAME_BY_ID,
  REQUEST_EDIT_CONTROL_BY_ID,
  CANCEL_EDIT_CONTROL_BY_ID,
  ADD_CONTROL_FRAME,
  AddControlFrameMutationResponseData,
  AddControlFrameMutationVariables,
  EditControlFrameTimeMutationResponseData,
  EditControlFrameTimeMutationVariables,
} from "@/graphql";

import type {
  ControlMapStatus,
  PosMapStatus,
  ControlRecord,
  ControlMapQueryPayload,
} from "@/core/models";
import { isControlMapStatus } from "@/core/models";

import { toControlMapStatusMutationPayload } from "@/core/utils/convert";

/**
 * controlAgent: responsible for controlMap and controlRecord
 */
export const controlAgent = {
  getControlMapPayload: async () => {
    const { data: controlMapData } = await client.query({
      query: GET_CONTROL_MAP,
    });

    return controlMapData.ControlMap.frameIds as ControlMapQueryPayload;
  },

  getControlRecord: async () => {
    const { data: controlRecordData } = await client.query({
      query: GET_CONTROL_RECORD,
    });

    return controlRecordData.controlFrameIDs as ControlRecord;
  },

  addFrame: async (addFrameInput: {
    start: number;
    frame: ControlMapStatus | PosMapStatus;
    fade?: boolean;
  }) => {
    if (!isControlMapStatus(addFrameInput.frame)) {
      return;
    }

    try {
      const { data: response } = await client.mutate<
        AddControlFrameMutationResponseData,
        AddControlFrameMutationVariables
      >({
        mutation: ADD_CONTROL_FRAME,
        variables: {
          start: addFrameInput.start,
          controlData: toControlMapStatusMutationPayload(addFrameInput.frame),
          ...(addFrameInput.fade && { fade: addFrameInput.fade }),
        },
      });

      return response?.addControlFrame;
    } catch (error) {
      console.error(error);
      throw error;
    }
  },

  saveFrame: async (saveFrameInput: {
    frameId: number;
    frame: ControlMapStatus | PosMapStatus;
    start: number;
    requestTimeChange: boolean;
    fade?: boolean;
  }) => {
    if (!isControlMapStatus(saveFrameInput.frame)) {
      return;
    }

    const promises: Promise<unknown>[] = [
      client.mutate({
        mutation: EDIT_CONTROL_FRAME,
        variables: {
          input: {
            frameId: saveFrameInput.frameId,
            controlData: toControlMapStatusMutationPayload(
              saveFrameInput.frame
            ),
            fade: saveFrameInput.fade,
          },
        },
      }),
    ];

    if (saveFrameInput.requestTimeChange) {
      promises.push(
        client.mutate<
          EditControlFrameTimeMutationResponseData,
          EditControlFrameTimeMutationVariables
        >({
          mutation: EDIT_CONTROL_FRAME_TIME,
          variables: {
            input: {
              frameID: saveFrameInput.frameId,
              start: saveFrameInput.start,
            },
          },
        })
      );
    }

    try {
      await Promise.all(promises);
    } catch (error) {
      console.error(error);
      throw error;
    }
  },

  deleteFrame: async (frameId: number) => {
    try {
      client.mutate({
        mutation: DELETE_CONTROL_FRAME_BY_ID,
        variables: {
          input: {
            frameID: frameId,
          },
        },
      });
    } catch (error) {
      console.error(error);
      throw error;
    }
  },

  requestEditPermission: async (frameId: number) => {
    const { data: response } = await client.mutate({
      mutation: REQUEST_EDIT_CONTROL_BY_ID,
      variables: {
        frameId: frameId,
      },
    });

    return response.RequestEditControl.ok;
  },

  cancelEditPermission: async (frameId: number) => {
    const { data: response } = await client.mutate({
      mutation: CANCEL_EDIT_CONTROL_BY_ID,
      variables: {
        frameId: frameId,
      },
    });

    return response.CancelEditControl.ok;
  },
};

File: editor-blender/api/control_agent.py
import asyncio
from collections.abc import Coroutine
from dataclasses import dataclass
from typing import Any

from ..client import client
from ..core.log import logger
from ..core.models import ColorID, ControlMap, ControlRecord, LEDEffectID, MapID
from ..core.utils.convert import control_map_query_to_state
from ..schemas.mutations import (
    ADD_CONTROL_FRAME,
    CANCEL_EDIT_CONTROL_BY_ID,
    DELETE_CONTROL_FRAME,
    EDIT_CONTROL_FRAME,
    EDIT_CONTROL_FRAME_TIME,
    REQUEST_EDIT_CONTROL_BY_ID,
    MutCancelEditControlResponse,
    MutDancerLEDStatusPayload,
    MutDancerStatusPayload,
    MutDeleteControlFrameInput,
    MutEditControlFrameInput,
    MutEditControlFrameTimeInput,
    MutRequestEditControlResponse,
)
from ..schemas.queries import (
    GET_CONTROL_MAP,
    GET_CONTROL_RECORD,
    QueryControlMapData,
    QueryControlMapPayload,
    QueryControlRecordData,
)


@dataclass
class ControlAgent:
    async def get_control_record(self) -> ControlRecord | None:
        """Get the control record from the server."""
        try:
            response = await client.execute(
                QueryControlRecordData,
                GET_CONTROL_RECORD,
            )
            controlRecord = response["controlFrameIDs"]

            return controlRecord

        except asyncio.CancelledError:
            pass

        except Exception:
            logger.exception("Failed to get control record")

        return None

    async def get_control_map_payload(self) -> QueryControlMapPayload | None:
        """Get the control map raw payload from the server"""
        try:
            response = await client.execute(QueryControlMapData, GET_CONTROL_MAP)
            controlMap = response["ControlMap"]

            return controlMap.frameIds

        except asyncio.CancelledError:
            pass

        except Exception:
            logger.exception("Failed to get control map payload")

        return None

    async def get_control_map(self) -> ControlMap | None:
        """Get the control map from the server."""
        try:
            response = await client.execute(QueryControlMapData, GET_CONTROL_MAP)
            controlMap = response["ControlMap"]

            return control_map_query_to_state(controlMap.frameIds)

        except asyncio.CancelledError:
            pass

        except Exception:
            logger.exception("Failed to get control map")

        return None

    async def add_frame(
        self,
        start: int,
        fade: bool,
        controlData: list[MutDancerStatusPayload],
        ledControlData: list[MutDancerLEDStatusPayload],
    ) -> str | None:
        """Add a new control frame to the control map."""
        try:
            response = await client.execute(
                str,
                ADD_CONTROL_FRAME,
                {
                    "start": start,
                    "controlData": controlData,
                    "fade": fade,
                    "ledControlData": ledControlData,
                },
            )
            return response["addControlFrame"]

        except asyncio.CancelledError:
            pass

        except Exception as err:
            logger.exception("Failed to add control frame")
            raise err

    # TODO: Support only change fade
    async def save_frame(
        self,
        id: MapID,
        controlData: list[MutDancerStatusPayload],
        ledControlData: list[MutDancerLEDStatusPayload],
        fade: bool | None = None,
        start: int | None = None,
    ):
        """Edit a control frame in the control map."""
        try:
            tasks: list[Coroutine[Any, Any, Any]] = []

            tasks.append(
                client.execute(
                    str,
                    EDIT_CONTROL_FRAME,
                    {
                        "input": MutEditControlFrameInput(
                            frameId=id,
                            controlData=controlData,
                            ledBulbData=ledControlData,
                            fade=fade,
                        )
                    },
                )
            )

            if start is not None:
                tasks.append(
                    client.execute(
                        str,
                        EDIT_CONTROL_FRAME_TIME,
                        {
                            "input": MutEditControlFrameTimeInput(
                                frameId=id, start=start
                            )
                        },
                    )
                )

            await asyncio.gather(*tasks)

        except asyncio.CancelledError:
            pass

        except Exception as err:
            logger.exception("Failed to save control frame")
            raise err

    async def delete_frame(self, id: MapID) -> str | None:
        """Delete a control frame from the control map."""
        try:
            response = await client.execute(
                str,
                DELETE_CONTROL_FRAME,
                {"input": MutDeleteControlFrameInput(frameID=id)},
            )
            return response["deleteControlFrame"]

        except asyncio.CancelledError:
            pass

        except Exception as err:
            logger.exception("Failed to delete control frame")
            raise err

    async def request_edit(self, id: MapID) -> bool | None:
        """Request to edit a control frame.
        Returns True if the request is successful, the server will prevent other users from editing the frame.
        Returns False if other users are editing the frame.
        """
        try:
            response = await client.execute(
                MutRequestEditControlResponse,
                REQUEST_EDIT_CONTROL_BY_ID,
                {"frameId": id},
            )
            return response["RequestEditControl"].ok

        except asyncio.CancelledError:
            pass

        except Exception as err:
            logger.exception("Failed to request edit control frame")
            raise err

    async def cancel_edit(self, id: MapID) -> bool | None:
        """Cancel the edit request of the control frame."""
        try:
            response = await client.execute(
                MutCancelEditControlResponse, CANCEL_EDIT_CONTROL_BY_ID, {"frameId": id}
            )
            return response["CancelEditControl"].ok

        except asyncio.CancelledError:
            pass

        except Exception as err:
            logger.exception("Failed to cancel edit control frame")
            raise err


control_agent = ControlAgent()
